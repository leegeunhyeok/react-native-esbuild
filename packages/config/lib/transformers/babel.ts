import deepmerge from 'deepmerge';
import type { TransformOptions } from '@babel/core';

export const getBabelOptions = (
  rootDir: string,
  customBabelOptions?: Partial<TransformOptions>,
): TransformOptions => {
  const baseOptions: TransformOptions = {
    minified: false,
    compact: false,
    sourceMaps: false,
    root: rootDir,
    highlightCode: !process.stdin.isTTY,
  };

  return customBabelOptions
    ? deepmerge(baseOptions, customBabelOptions)
    : baseOptions;
};
