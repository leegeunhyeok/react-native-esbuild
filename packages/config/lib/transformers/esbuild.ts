import path from 'node:path';
import deepmerge from 'deepmerge';
import type { BuildOptions } from 'esbuild';
import {
  DEFAULT_ENTRY_POINT,
  DEFAULT_OUTFILE,
  SOURCE_EXTENSIONS,
  ASSET_EXTENSIONS,
} from '../shares';
import type { BundleConfig } from '../types';

export function getEsbuildOptions(
  bundleConfig: BundleConfig,
  customEsbuildOptions?: Partial<BuildOptions>,
): BuildOptions {
  const {
    entry = DEFAULT_ENTRY_POINT,
    outfile = DEFAULT_OUTFILE,
    platform,
    minify,
  } = bundleConfig;

  const platforms = [platform, 'native', 'react-native'];
  const extensions = SOURCE_EXTENSIONS.concat(ASSET_EXTENSIONS);

  const baseOptions: BuildOptions = {
    /**
     * mainFields
     * @see {@link https://github.com/facebook/metro/blob/0.72.x/docs/Configuration.md#resolvermainfields}
     */
    mainFields: ['react-native', 'browser', 'main', 'module'],
    entryPoints: [entry],
    outfile,
    sourceRoot: path.dirname(entry),
    resolveExtensions: platforms
      .map((platform) => extensions.map((ext) => `.${platform}${ext}`))
      .concat(extensions)
      .flat(),
    loader: Object.fromEntries(ASSET_EXTENSIONS.map((ext) => [ext, 'file'])),
    legalComments: 'none',
    target: 'es6',
    supported: {
      /**
       * to avoid block scope bug on hermes engine.
       * in `__copyProps`, generated by esbuild
       *
       * @see {@link https://github.com/evanw/esbuild/blob/v0.18.17/internal/runtime/runtime.go#L192}
       */
      'for-of': false,
    },
    format: 'esm',
    bundle: true,
    sourcemap: true,
    minify,
    logLevel: 'silent',
  };

  return deepmerge(baseOptions, customEsbuildOptions ?? {});
}
