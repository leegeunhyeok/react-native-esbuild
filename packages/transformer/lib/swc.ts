import {
  transform,
  minify,
  type Options,
  type JsMinifyOptions,
} from '@swc/core';
import { getSwcOptions } from '@react-native-esbuild/config';
import type { Transformer } from './types';

export const transformWithSwc: Transformer<Options> = async (
  code,
  context,
  customOptions,
) => {
  const options = getSwcOptions(
    {
      filename: context.path,
      root: context.root,
    },
    customOptions,
  );
  const { code: transformedCode } = await transform(code, options);

  if (typeof transformedCode !== 'string') {
    throw new Error('swc transformed source is empty');
  }

  return transformedCode;
};

export const minifyWithSwc: Transformer<JsMinifyOptions> = async (
  code,
  _context,
  customOptions,
) => {
  const { code: minifiedCode } = await minify(code, customOptions);

  if (typeof minifiedCode !== 'string') {
    throw new Error('swc minified source is empty');
  }

  return minifiedCode;
};
