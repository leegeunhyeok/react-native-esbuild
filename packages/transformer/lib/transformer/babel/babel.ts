import type { TransformOptions } from '@babel/core';
import { loadOptions, transformAsync, transformSync } from '@babel/core';
import type {
  Transformer,
  SyncTransformer,
  TransformerContext,
} from '../../types';

const loadBabelOptions = (
  context: TransformerContext,
  options?: TransformOptions,
): ReturnType<typeof loadOptions> => {
  return loadOptions({
    ...options,
    root: context.root,
    filename: context.path,
  });
};

export const transformWithBabel: Transformer<TransformOptions> = async (
  code: string,
  context,
  options,
) => {
  const babelOptions = loadBabelOptions(context, options);
  if (!babelOptions) {
    throw new Error('cannot load babel options');
  }

  const result = await transformAsync(code, babelOptions);
  if (typeof result?.code !== 'string') {
    throw new Error('babel transformed source is empty');
  }

  return result.code;
};

export const transformSyncWithBabel: SyncTransformer<TransformOptions> = (
  code: string,
  context,
  options,
) => {
  const babelOptions = loadBabelOptions(context, options);
  if (!babelOptions) {
    throw new Error('cannot load babel options');
  }

  const result = transformSync(code, babelOptions);
  if (typeof result?.code !== 'string') {
    throw new Error('babel transformed source is empty');
  }

  return result.code;
};
