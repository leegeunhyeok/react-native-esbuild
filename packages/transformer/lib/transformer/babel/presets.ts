import type { TransformOptions } from '@babel/core';

const getCommon = (): TransformOptions => ({
  minified: false,
  compact: false,
  sourceMaps: false,
  babelrc: false,
  highlightCode: !process.stdin.isTTY,
});

export { getCommon };
