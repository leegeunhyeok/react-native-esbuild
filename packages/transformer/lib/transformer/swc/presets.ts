import type { Options } from '@swc/core';

/**
 * swc transform options preset for react-native runtime.
 */
const getReactNativeRuntimeOptions = (): Options => ({
  minify: false,
  sourceMaps: false,
  isModule: true,
  inputSourceMap: false,
  inlineSourcesContent: false,
  jsc: {
    target: 'es5',
    loose: false,
    externalHelpers: true,
    keepClassNames: true,
  },
});

/**
 * swc transform options preset for jest.
 */
const getJestOptions = (): Options => ({
  sourceMaps: 'inline',
  jsc: {
    target: 'es2022',
    transform: {
      /**
       * @see {@link https://github.com/swc-project/jest/blob/v0.2.29/index.ts#L119}
       */
      // @ts-expect-error -- swc sugar code
      hidden: {
        jest: true,
      },
      react: {
        runtime: 'automatic',
      },
    },
  },
  module: { type: 'commonjs' },
});

export { getReactNativeRuntimeOptions, getJestOptions };
