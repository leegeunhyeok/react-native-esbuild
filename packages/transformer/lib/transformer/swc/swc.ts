import {
  transform,
  transformSync,
  minify,
  type Options,
  type JsMinifyOptions,
  type TsParserConfig,
  type EsParserConfig,
} from '@swc/core';
import type {
  Transformer,
  SyncTransformer,
  TransformerContext,
} from '../../types';

const getParserOptions = (path: string): TsParserConfig | EsParserConfig => {
  return /\.tsx?$/.test(path)
    ? ({
        syntax: 'typescript',
        tsx: true,
        dynamicImport: true,
      } as TsParserConfig)
    : ({
        syntax: 'ecmascript',
        jsx: true,
        exportDefaultFrom: true,
      } as EsParserConfig);
};

const getSwcOptions = (
  context: TransformerContext,
  options?: Options,
): Options => {
  return {
    ...options,
    jsc: {
      parser: getParserOptions(context.path),
      ...options?.jsc,
    },
    filename: context.path,
    root: context.root,
  };
};

export const transformWithSwc: Transformer<Options> = async (
  code,
  context,
  options,
) => {
  const { code: transformedCode } = await transform(
    code,
    getSwcOptions(context, options),
  );

  if (typeof transformedCode !== 'string') {
    throw new Error('swc transformed source is empty');
  }

  return transformedCode;
};

export const transformSyncWithSwc: SyncTransformer<Options> = (
  code,
  context,
  options,
) => {
  const { code: transformedCode } = transformSync(
    code,
    getSwcOptions(context, options),
  );

  if (typeof transformedCode !== 'string') {
    throw new Error('swc transformed source is empty');
  }

  return transformedCode;
};

export const minifyWithSwc: Transformer<JsMinifyOptions> = async (
  code,
  _context,
  options,
) => {
  const { code: minifiedCode } = await minify(code, options);

  if (typeof minifiedCode !== 'string') {
    throw new Error('swc minified source is empty');
  }

  return minifiedCode;
};
