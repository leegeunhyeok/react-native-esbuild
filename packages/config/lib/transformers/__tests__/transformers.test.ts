import { getEsbuildOptions } from '../esbuild';
import { getSwcOptions } from '../swc';
import { getBabelOptions } from '../babel';

describe('getEsbuildOptions', () => {
  it('should match snapshot', () => {
    expect(
      getEsbuildOptions({
        platform: 'android',
        entry: 'index.js',
        outfile: 'main.bundlejs',
        dev: true,
        minify: false,
        metafile: false,
        assetsDir: 'assets',
      }),
    ).toMatchSnapshot();
  });
});

describe('getSwcOptions', () => {
  it('should match snapshot', () => {
    expect(
      getSwcOptions({ filename: 'file.js', root: '/root' }),
    ).toMatchSnapshot();
  });

  describe('when filename ends with `.js`', () => {
    it('should set parse syntax to `ecmascript`', () => {
      const options = getSwcOptions({ filename: 'file.js', root: '/root' });
      expect(options.jsc?.parser?.syntax).toEqual('ecmascript');
    });
  });

  describe('when filename ends with `.jsx`', () => {
    it('should set parse syntax to `ecmascript`', () => {
      const options = getSwcOptions({ filename: 'file.jsx', root: '/root' });
      expect(options.jsc?.parser?.syntax).toEqual('ecmascript');
    });
  });

  describe('when filename ends with `.ts`', () => {
    it('should set parse syntax to `typescript`', () => {
      const options = getSwcOptions({ filename: 'file.ts', root: '/root' });
      expect(options.jsc?.parser?.syntax).toEqual('typescript');
    });
  });

  describe('when filename ends with `.tsx`', () => {
    it('should set parse syntax to `typescript`', () => {
      const options = getSwcOptions({ filename: 'file.tsx', root: '/root' });
      expect(options.jsc?.parser?.syntax).toEqual('typescript');
    });
  });

  describe('when customSwcOptions present', () => {
    it('should override default options', () => {
      const overrideOptions = {
        isModule: false,
        jsc: {
          parser: undefined,
          loose: false,
        },
      };
      const options = getSwcOptions(
        { filename: 'file.js', root: '/root' },
        overrideOptions,
      );

      expect(options).toMatchObject(
        expect.objectContaining({
          ...overrideOptions,
          jsc: expect.objectContaining(overrideOptions.jsc),
        }),
      );
    });
  });
});

describe('getBabelOptions', () => {
  it('should match snapshot', () => {
    expect(getBabelOptions('/root')).toMatchSnapshot();
  });

  describe('when customBabelOptions present', () => {
    it('should override default options', () => {
      const overrideOptions = {
        minified: true,
        compact: true,
        sourceMaps: true,
      };
      const options = getBabelOptions('/root', overrideOptions);

      expect(options).toMatchObject(expect.objectContaining(overrideOptions));
    });
  });
});
