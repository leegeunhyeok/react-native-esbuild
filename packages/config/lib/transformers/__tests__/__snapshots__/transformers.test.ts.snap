// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBabelOptions should match snapshot 1`] = `
{
  "compact": false,
  "minified": false,
  "root": "/root",
  "sourceMaps": false,
}
`;

exports[`getEsbuildOptions should match snapshot 1`] = `
{
  "bundle": true,
  "entryPoints": [
    "index.js",
  ],
  "format": "esm",
  "legalComments": "none",
  "loader": {
    ".3g2": "file",
    ".3gp": "file",
    ".aac": "file",
    ".avi": "file",
    ".bmp": "file",
    ".eot": "file",
    ".gif": "file",
    ".ico": "file",
    ".jpeg": "file",
    ".jpg": "file",
    ".mid": "file",
    ".midi": "file",
    ".mp3": "file",
    ".mp4": "file",
    ".mpeg": "file",
    ".mpg": "file",
    ".oga": "file",
    ".ogv": "file",
    ".otf": "file",
    ".png": "file",
    ".svg": "file",
    ".tif": "file",
    ".tiff": "file",
    ".ttf": "file",
    ".wav": "file",
    ".webm": "file",
    ".webp": "file",
    ".woff": "file",
    ".woff2": "file",
  },
  "logLevel": "silent",
  "mainFields": [
    "react-native",
    "browser",
    "main",
    "module",
  ],
  "metafile": false,
  "minify": false,
  "outfile": "main.bundlejs",
  "resolveExtensions": [
    ".android.tsx",
    ".android.ts",
    ".android.jsx",
    ".android.js",
    ".android.mjs",
    ".android.cjs",
    ".android.json",
    ".android.avi",
    ".android.mp4",
    ".android.mpeg",
    ".android.mpg",
    ".android.ogv",
    ".android.webm",
    ".android.3gp",
    ".android.3g2",
    ".android.aac",
    ".android.midi",
    ".android.mid",
    ".android.mp3",
    ".android.oga",
    ".android.wav",
    ".android.3gp",
    ".android.3g2",
    ".android.eot",
    ".android.otf",
    ".android.ttf",
    ".android.woff",
    ".android.woff2",
    ".android.bmp",
    ".android.gif",
    ".android.ico",
    ".android.jpeg",
    ".android.jpg",
    ".android.png",
    ".android.svg",
    ".android.tiff",
    ".android.tif",
    ".android.webp",
    ".native.tsx",
    ".native.ts",
    ".native.jsx",
    ".native.js",
    ".native.mjs",
    ".native.cjs",
    ".native.json",
    ".native.avi",
    ".native.mp4",
    ".native.mpeg",
    ".native.mpg",
    ".native.ogv",
    ".native.webm",
    ".native.3gp",
    ".native.3g2",
    ".native.aac",
    ".native.midi",
    ".native.mid",
    ".native.mp3",
    ".native.oga",
    ".native.wav",
    ".native.3gp",
    ".native.3g2",
    ".native.eot",
    ".native.otf",
    ".native.ttf",
    ".native.woff",
    ".native.woff2",
    ".native.bmp",
    ".native.gif",
    ".native.ico",
    ".native.jpeg",
    ".native.jpg",
    ".native.png",
    ".native.svg",
    ".native.tiff",
    ".native.tif",
    ".native.webp",
    ".react-native.tsx",
    ".react-native.ts",
    ".react-native.jsx",
    ".react-native.js",
    ".react-native.mjs",
    ".react-native.cjs",
    ".react-native.json",
    ".react-native.avi",
    ".react-native.mp4",
    ".react-native.mpeg",
    ".react-native.mpg",
    ".react-native.ogv",
    ".react-native.webm",
    ".react-native.3gp",
    ".react-native.3g2",
    ".react-native.aac",
    ".react-native.midi",
    ".react-native.mid",
    ".react-native.mp3",
    ".react-native.oga",
    ".react-native.wav",
    ".react-native.3gp",
    ".react-native.3g2",
    ".react-native.eot",
    ".react-native.otf",
    ".react-native.ttf",
    ".react-native.woff",
    ".react-native.woff2",
    ".react-native.bmp",
    ".react-native.gif",
    ".react-native.ico",
    ".react-native.jpeg",
    ".react-native.jpg",
    ".react-native.png",
    ".react-native.svg",
    ".react-native.tiff",
    ".react-native.tif",
    ".react-native.webp",
    ".tsx",
    ".ts",
    ".jsx",
    ".js",
    ".mjs",
    ".cjs",
    ".json",
    ".avi",
    ".mp4",
    ".mpeg",
    ".mpg",
    ".ogv",
    ".webm",
    ".3gp",
    ".3g2",
    ".aac",
    ".midi",
    ".mid",
    ".mp3",
    ".oga",
    ".wav",
    ".3gp",
    ".3g2",
    ".eot",
    ".otf",
    ".ttf",
    ".woff",
    ".woff2",
    ".bmp",
    ".gif",
    ".ico",
    ".jpeg",
    ".jpg",
    ".png",
    ".svg",
    ".tiff",
    ".tif",
    ".webp",
  ],
  "sourceRoot": ".",
  "sourcemap": true,
  "supported": {
    "for-of": false,
  },
  "target": "es6",
}
`;

exports[`getSwcOptions should match snapshot 1`] = `
{
  "inlineSourcesContent": false,
  "inputSourceMap": false,
  "isModule": true,
  "jsc": {
    "externalHelpers": true,
    "keepClassNames": true,
    "loose": false,
    "parser": {
      "exportDefaultFrom": true,
      "jsx": true,
      "syntax": "ecmascript",
    },
    "target": "es5",
  },
  "minify": false,
  "root": "/root",
  "sourceMaps": false,
}
`;
